snippets/base/management/commands/fetch_daily_metrics.py

```diff
@@ -0,0 +1,109 @@
+        )
+
+    def handle(self, *args, **options):
+        if not options['date']:
+            now = datetime.utcnow()
+            date = (
+                datetime(year=now.year, month=now.month, day=now.day, hour=0, minute=0, second=0) -
+                timedelta(days=1)
+            )
+        else:
+            date = datetime.strptime(options['date'], '%Y%m%d')
+
+        last_week = date - timedelta(days=7)  # date - 7 since we 're fetching data for yesterday.
+        date_next_day = date + timedelta(days=1)
+
+        if not settings.REDASH_API_KEY:
+            raise CommandError('Enviroment variable REDASH_API_KEY is required.')
+
+        redash = RedashDynamicQuery(
+            endpoint=settings.REDASH_ENDPOINT,
+            apikey=settings.REDASH_API_KEY,
+            max_wait=settings.REDASH_MAX_WAIT,
+        )
+        jobs = (
+            Job.objects
+            .filter(status__in=[Job.PUBLISHED, Job.COMPLETED])
+            .filter(
+                # Publish start before the end of day of interest and still
+                # not completed.
+                (Q(publish_start__lt=date_next_day) & Q(completed_on=None)) |
+                # Or completed during the last 7 days before EOD of the day of interest.
+                (Q(completed_on__gte=last_week) & Q(completed_on__lt=date_next_day))
+            )
+        )
+
+        self.stdout.write(f'Fetching data for {date} for {jobs.count()} jobs.')
+
+        data_fetched = False
+        for job in jobs:
+            bind_data = {
+                'start_date': date.strftime('%Y%m%d'),
+                'end_date': date.strftime('%Y%m%d'),
+                'message_id': job.id,
+            }
+            try:
+                result = redash.query(settings.REDASH_QUERY_ID, bind_data)
+            except Exception as exp:
+                # Capture the exception but don't quit
+                sentry_sdk.capture_exception(exp)
+                continue
+
+            impressions = 0
+            clicks = 0
+            blocks = 0
+            try:
+                for row in result['query_result']['data']['rows']:
+                    if row['event'] == 'IMPRESSION':
+                        impressions = row['counts']
+                    elif row['event'] == 'BLOCK':
+                        blocks = row['counts']
+                    elif row['event'] in ['CLICK', 'CLICK_BUTTON']:
+                        clicks += row['counts']
+            except KeyError as exp:
+                # Capture the exception but don't quit
+                sentry_sdk.capture_exception(exp)
+                continue
+            else:
+                data_fetched = True
+
+            if not any([impressions, blocks, clicks]):
+                # Job has no metrics for date, ignore
+                continue
+
+            # Use update to avoid triggering Django signals and updating Job's
+            # and ASRSnippet's modified date.
+            DailyJobMetrics.objects.create(
```

Comment:
I think we may want to use `update_or_create` here to avoid duplicates that could result from running this command more than once for the same day.

Reply:
We shouldn't be running this command more than once per day but it shouldn't hurt to add update_or_create here.